pkg/storegateway/gateway_ring.go

```diff
@@ -109,6 +111,8 @@ func (cfg *RingConfig) RegisterFlags(f *flag.FlagSet) {
 	f.DurationVar(&cfg.WaitStabilityMinDuration, ringFlagsPrefix+"wait-stability-min-duration", time.Minute, "Minimum time to wait for ring stability at startup. 0 to disable.")
 	f.DurationVar(&cfg.WaitStabilityMaxDuration, ringFlagsPrefix+"wait-stability-max-duration", 5*time.Minute, "Maximum time to wait for ring stability at startup. If the store-gateway ring keeps changing after this period of time, the store-gateway will start anyway.")
 
+	f.DurationVar(&cfg.FinalSleep, ringFlagsPrefix+"final-sleep", 0*time.Second, "The sleep seconds when store-gateway is shutting down.")
```

Comment:
Why this has default value 0s but other components have 30s?

Reply:
because SG we have replications, even we don't sleep and other pods still think this SG is good, it is okay. We will only get one error.

Reply:
We cannot make such assumptions. Store gateway have non replicated mode as well.

Reply:
okay, sure. Set the default to zero for SG, Ruler, and Alertmanager, so it is upto the user to make the right decision for the final delay.

If they don't have HA for ruler,alertmanager and decided to use DDB KV, then the delay need to be larger than DDB puller-sync-time.

If they have HA for ruler,alertmanger, then this finalSleep is not needed. They can afford one error.

Reply:
We need to document this properly if we add this flag

Reply:
Where should I put the explanation. Kind of long. Tried to add some comments on the flag description. But don't think that is the good place to put long explanation.

Reply:
My recommendation
"The sleep seconds wait before pod shutdown allowing a delay to propagate ring changes."

@wenxu1024 we need to add this flag in the changelog. There you can explain better. Maybe that is what @yeya24 was referring to.

Reply:
You can document at https://cortexmetrics.io/docs/configuration/v1guarantees/